File: src/visual-graph/builder/visible-ancestor-subgraph.ts
import type { SerializedScope } from "../../ir/serialized/serialized-scope.js";
import type { VisualSubgraph } from "../visual-subgraph.js";
import type { BuildState } from "./build-state.js";
import type { BuilderContext } from "./context.js";

// Walk up scope.upper until we hit a scope that materialised a subgraph
// during the top-down build and return it. Collapsed ancestors are
// skipped (they have no subgraph). Returns null when the chain reaches
// the module / global root without ever crossing a surviving subgraph --
// callers treat null as "no visible host", which turns the corresponding
// edge into a drop signal.
export function visibleAncestorSubgraph(
  scope: SerializedScope,
  ctx: BuilderContext,
  state: BuildState,
): VisualSubgraph | null {
  let parentId = scope.upper;
  while (parentId !== null) {
    const sg = state.subgraphByScope.get(parentId);
    if (sg) {
      return sg;
    }
    const parent = ctx.scopeMap.get(parentId);
    if (!parent) {
      return null;
    }
    parentId = parent.upper;
  }
  return null;
}

File: crates/unsnarl-visual-graph/src/builder/visible_ancestor_subgraph.rs
//! Walk `scope.upper` upward until a surviving subgraph is found
//! and return its handle. Collapsed ancestors are skipped (they
//! never recorded a subgraph in `state.subgraph_by_scope`). Returns
//! `None` when the chain reaches the module / global root without
//! crossing a visible subgraph — callers treat that as a drop
//! signal for the corresponding edge.

use unsnarl_ir::serialized::SerializedScope;

use super::arena::SubgraphIdx;
use super::context::BuilderContext;
use super::state::BuildState;

pub fn visible_ancestor_subgraph(
    scope: &SerializedScope,
    ctx: &BuilderContext<'_>,
    state: &BuildState,
) -> Option<SubgraphIdx> {
    let mut parent_id = scope.upper.as_ref().map(|id| id.value());
    while let Some(id) = parent_id {
        if let Some(&idx) = state.subgraph_by_scope.get(id) {
            return Some(idx);
        }
        let parent = ctx.scope_map.get(id).copied()?;
        parent_id = parent.upper.as_ref().map(|i| i.value());
    }
    None
}
